### PR TITLE
fix(ingest): sort by last modified not working in the UI

### PR DESCRIPTION
**Pull request by sid-acryl**

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        1. `lastModified` of datasetProperties is not set
+        2. And there are operation aspects
+    in this case set the `lastModified` of datasetProperties to max value of operation aspects `lastUpdatedTimestamp`.
+
+    We need this functionality to support sort by `last modified` on UI.
+    """
+    candidate_dataset_for_patch: Dict[str, TimestampPair] = {}
+
+    for wu in stream:
+        if not wu.get_urn().startswith("urn:li:dataset:"):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:33]: use `guess_entity_type(...) == 'dataset'`

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    DatasetPropertiesClass,
+    OperationClass,
+    TimeStampClass,
+)
+from datahub.specific.dataset import DatasetPatchBuilder
+
+
+@dataclasses.dataclass
+class TimestampPair:
+    last_modified: Optional[TimeStampClass]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:15]: is this last modified from dataset props 

can we make the variable naming more clear?

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                and dataset_operation_aspect.lastUpdatedTimestamp
+            ):
+                if timestamp_pair.last_updated_timestamp:
+                    # lastUpdatedTimestamp is present and less than the current operation aspect lastUpdateTimestamp
+                    # than update the timestamp_pair
+                    if (
+                        timestamp_pair.last_updated_timestamp
+                        < dataset_operation_aspect.lastUpdatedTimestamp
+                    ):
+                        timestamp_pair.last_updated_timestamp = (
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:58]: use `pair.last_updated_timestamp = max(pair. last_updated_timestamp or 0, operation.field)`

that will let you avoid much of this if...else complexity

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:1]: this needs unit tests in `test_source_helpers.py`

**Review by hsheth2 — CHANGES_REQUESTED** [review]
What happens if this is used with a source that supports `patch_custom_properties`? Please add some tests for the scenario where the source produces MCP patches for dataset properties

e.g. redshift does this

https://github.com/datahub-project/datahub/blob/6704d444704c11239b6a3ee0075fd5045c3e2f89/metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py#L834

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py (hunk 2 above)
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:15]: I can rename it to `last_modified_dataset_properties_aspect` but it is too long, so I have added comment next to variable names.

**Comment by sid-acryl** [comment]
> What happens if this is used with a source that supports `patch_custom_properties`? Please add some tests for the scenario where the source produces MCP patches for dataset properties
> 
> e.g. redshift does this
> 
> https://github.com/datahub-project/datahub/blob/6704d444704c11239b6a3ee0075fd5045c3e2f89/metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py#L834


In this case the final work-units contains two patch work units, one from `source` and second from `auto_patch_last_modified`

File: metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py [hunk 4]
```diff
@@ -656,3 +661,79 @@ def test_auto_empty_dataset_usage_statistics_invalid_timestamp(
...
+
+    initial_wu = list(auto_workunit(mcps))
+
+    expected = initial_wu
+
+    assert (
+        list(auto_patch_last_modified(initial_wu)) == expected
+    )  # There should be no change
+
+    mcps.append(
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py:683]: it seems like these can be split up into multiple separate tests

File: metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py [hunk 5]
```diff
@@ -656,3 +661,79 @@ def test_auto_empty_dataset_usage_statistics_invalid_timestamp(
...
+                operationType=OperationTypeClass.CREATE,
+            ),
+        )
+    )
+
+    expected = list(auto_workunit(mcps))
+
+    dataset_patch_builder = DatasetPatchBuilder(
+        urn="urn:li:dataset:a.b.c"
+    ).set_last_modified(TimeStampClass(time=20))
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py:709]: need a test where the inputs into the helper has a MCP generated by DatasetPatchBuilder

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    OperationClass,
+    TimeStampClass,
+)
+from datahub.specific.dataset import DatasetPatchBuilder
+from datahub.utilities.urns.urn import guess_entity_type
+
+
+@dataclasses.dataclass
+class TimestampPair:
+    last_modified: Optional[TimeStampClass]  # last_modified of datasetProperties aspect
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py:16]: comments are a crutch to use if the code isn't clear. But if we can make the variable names clear and not need comments, that's better

please update the variable names to be more clear

File: metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py [hunk 7]
```diff
@@ -656,3 +661,79 @@ def test_auto_empty_dataset_usage_statistics_invalid_timestamp(
...
+                timestampMillis=10,
+                lastUpdatedTimestamp=12,
+                operationType=OperationTypeClass.CREATE,
+            ),
+        )
+    )
+
+    mcps.append(
+        MetadataChangeProposalWrapper(
+            entityUrn="urn:li:dataset:a.b.c",
```
Comment (hsheth2) [inline metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py:696]: this isn't a valid urn

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py (hunk 7 above)
Comment (sid-acryl) [reply metadata-ingestion/tests/unit/api/source_helpers/test_source_helpers.py:696]: done

File: metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/api/auto_work_units/auto_dataset_properties_aspect.py]: surely you can come up with a variable name that is both shorter than that and more clear than `last_modified`

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**